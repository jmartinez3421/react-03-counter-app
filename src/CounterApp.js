//Para usar el useState tenemos que importarlo
import React, {useState} from 'react';
import PropTypes from 'prop-types';

const CounterApp = ({value}) => {

    //Usaremos el método useState haciendo una deconstruccion del array que lo compone, obteniendo primero el valor y
    //la segunda posición que hace referencia a la función que usaremos para cambiar el valor del state
    const [ counter, setCounter ]  = useState( value );

    const handleAdd = () => {
        setCounter(counter + 1);
    }

    const handleReset = () => {
        setCounter(value);
    }

    const handleSubstract = () => {
        setCounter(counter - 1);
    }

    return (
        <>
            <h1>CounterApp</h1>
            <h2> {counter} </h2>

            {/* 
               Si al pasarle la función al onclick la ponemos como handleAdd() React la va a ejecutar cuando renderice
                el botón dejandolo inservible, con lo que cuando pasemos las funciones debemos ponerlas sin ().
                Solo las pondremos con () en el caso de que la función que tenemos que lanzar nos devuelva otra función.
            */}
            <button onClick={ handleAdd }>+1</button>
            <button onClick={ handleReset }>Reset</button>
            <button onClick={ handleSubstract }>-1</button>
        </>
    );
}

CounterApp.propTypes = {
    value: PropTypes.number,
};

CounterApp.defaultProps = {
    value: 5,
}

export default CounterApp;
