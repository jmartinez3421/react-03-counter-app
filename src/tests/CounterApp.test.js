import "@testing-library/jest-dom";
import CounterApp from '../CounterApp';
import React from "react";
import { shallow } from "enzyme";

describe('Pruebas sobre el componente CounterApp', () => {

    //Inicialimzamos el wrapper para no tener que definirlo en cada test
    let wrapper = shallow(<CounterApp />);

    //Con el método beforeEach ejecutaremos la funcion que definimos dentro antes de ejecutar cada test
    beforeEach(() => {

        //Reiniciamos la variable wrapper para que no tenga el valor alterado
        wrapper = shallow(<CounterApp />);

    });

    test('el componente CounterApp debe imprimir correctamente el contenido', () => {

        expect(wrapper).toMatchSnapshot();

    });

    test('el valor por defecto del contador tiene que ser igual al que pasamos por props', () => {

        const value = 100;
        const wrapper = shallow(<CounterApp value={value} />);

        //Ponemos .trim() para eliminar los espacios del texto
        const valorCounter = wrapper.find('h2').text().trim();

        //Ponemos el value de esta manera porque tiene que ser como String ya que así obtenemos el valor del counter
        expect(valorCounter).toBe(`${value}`);
    });

    test('debe incrementar el valor del contador', () => {

        //Hacemos una referencia al botón de incrementar, que es el que se encuentra en primera posición
        const boton1 = wrapper.find('button').at(0);
        //Simulamos el evento click usando el método simulate
        boton1.simulate('click');

        //Obtenemos el valor del counter una vez lanzado el evento click
        const valorCounter = wrapper.find('h2').text().trim();

        //Comprobamos si se ha incrementado el valor poniendo que se espera que sea 6 (por defecto es 5)
        expect(valorCounter).toBe('6');
    });

    test('debe restar 1 al valor del contador', () => {

        const boton1 = wrapper.find('button').at(2);
        boton1.simulate('click');

        const valorCounter = wrapper.find('h2').text().trim();

        expect(valorCounter).toBe('4');
    });

    test('debe restaurar el valor por defecto de counter, el cual es 5', () => {
        
        //Aumentamos el valor del counter simulamos 
        wrapper.find('button').at(0).simulate('click');
        console.log(wrapper.find('h2').text().trim());

        const boton1 = wrapper.find('button').at(1);
        boton1.simulate('click');

        const valorCounter = wrapper.find('h2').text().trim();

        expect(valorCounter).toBe('5');
    })
})