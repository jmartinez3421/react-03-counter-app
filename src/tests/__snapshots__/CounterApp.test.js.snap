// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas sobre el componente CounterApp el componente CounterApp debe imprimir correctamente el contenido 1`] = `
<Fragment>
  <h1>
    CounterApp
  </h1>
  <h2>
     
    5
     
  </h2>
  <button
    onClick={[Function]}
  >
    +1
  </button>
  <button
    onClick={[Function]}
  >
    Reset
  </button>
  <button
    onClick={[Function]}
  >
    -1
  </button>
</Fragment>
`;
